scripts: kconfig: Add dt_compat_get_inst_prop method

nandojve opened the pull request:
The **dt_str_val** method has been deprecated. This add an alternative
that uses **compat** and **instance number** to find a node and than
return the string representation or a passed default value.

This is necessary becase **path** is valid only at SoC directory. For
instance, a driver for a SPI bus need search for a **compat** based on
the **instance** to determine the proper configuration.

depends on: #22215

Example: how determine the right interface name from the device tree
node label.

```
at DT:

&foo0 {
	compatible = "manufacturer,device"
	status = "disabled";
	label = "FOO_0";
};
&foo1 {
	compatible = "manufacturer,device"
	status = "okay";
	label = "FOO_1";
};

at Kconfig:

DT_FOO      := manufacturer,device
DT_FOO_INST := 0
DT_FOO_DEF  := "default string on not match/error - optional empty"

config FOO
	string "text"
	default "$(dt_compat_get_str, \
		 label, \
		 $(DT_FOO), \
		 $(DT_FOO_INST), \
		 $(DT_FOO_DEF))"

The result is CONFIG_FOO="FOO_1"
```
An example of use was delivered too.

nandojve committed:
scripts: kconfig: Add dt_compat_get_inst_prop method

The 'dt_str_val' method has been deprecated. This add an alternative
that uses 'compatible' and 'instance index' to find a node and than
return the string representation or a passed default value. The
instance is indexed by occurences of a compatible node with property
'status=okay'.

This is necessary because 'path' is valid only at SoC directory. For
instance, a driver for a SPI bus need search for a 'compatible' based
on the 'instance' to determine the proper configuration.

Example: how determine the right interface name from the device tree
node label for the first instance?

at DT:

&foo0 {
	compatible = "manufacturer,device"
	status = "disabled";
	label = "FOO_0";
};
&foo1 {
	compatible = "manufacturer,device"
	status = "okay";
	label = "FOO_1";
};

at Kconfig:

DT_FOO      := manufacturer,device
DT_FOO_INST := 0
DT_FOO_DEF  := "default string on not match/error - optional empty"

config FOO
	string "text"
	# Fetch 'label' property from instance INST, defaulting to DEF
	default "$(dt_compat_get_inst_prop, \
		 label, \
		 $(DT_FOO), \
		 $(DT_FOO_INST), \
		 $(DT_FOO_DEF))"

The result is CONFIG_FOO="FOO_1".

'dt_compat_get_inst_prop' an be used to fetch properties that store
string or integer values.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: ieee802154: rf2xx: Get default dev name from DT

The IEEE 802.15.4 examples uses NET_CONFIG_IEEE802154_DEV_NAME config
to determine the default device name to bind. This allows user define
at DT node label the iface name. The default for all examples will be
the first interface instance. This ensure on a multi band board with
2.4G and Sub-Giga radios that the first defined radio will be used.
User can switch to the second radio definition by simply disabling the
first one.

Since this configuration is applied at Kconfig, changes obligate user
to rebuild all the application. This ensure that cache will not affect
expected behaviour.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
Rebase and Fix 364 line.
```
        if prop not in node.props:
            continue
```
to
```
        if prop not in node.props:
            break
```

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,47 @@ def dt_compat_enabled(kconf, _, compat):
+    property by the name of 'prop'. If the 'prop' exists it will return the string
+    representation of that 'prop' otherwise return ret_def.
+    """
+
```
No blank line after help text.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,47 @@ def dt_compat_enabled(kconf, _, compat):
+    representation of that 'prop' otherwise return ret_def.
+    """
+
+    ret_def_trimed = ret_def.strip()
```
*trimmed

nandojve replied:
Fixed with variables **\<parameter\>_s**. I noted that if user break lines at _Kconfig_ will be necessary execute **strip()** to avoid miss behaviors on all strings. This must be applied on all existing methods when applicable.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,47 @@ def dt_compat_enabled(kconf, _, compat):
+        return ret_def_trimed
+
+    if doc_mode or edt is None:
+        return ret_def_trimed
```
Could put this check above the `try` maybe.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,47 @@ def dt_compat_enabled(kconf, _, compat):
+        if prop not in node.props:
+            break
+
+        if node.props[prop].val:
```
I wonder if this check is needed. Maybe it could just be `return str(node.props[prop])`. If someone does e.g. `string-prop = ""`, then the empty string (which is falsy) is probably intentional.

nandojve replied:
You are right! I did a lot of tests to understand the structure and missed this.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,47 @@ def dt_compat_enabled(kconf, _, compat):
     return "n"
 
 
+def dt_compat_get_string(kconf, _, compat, inst=0, ret_def='', prop='label'):
```
Call it `get_str` instead of `get_string`, for consistency with the other functions. I wonder if `dt_str_from_compat` might be nice too. Then there could be a `dt_str_from_path` later on if needed.

A signature like `(kconf, _, prop, compat, inst, default="")` might be better, where you're forced to pass everything but the default value. Makes it a bit more explicit.

Use double quotes instead of single quotes, just for consistency with the rest of the code (I know it's less common in Python).

Could call it `default` instead of `ret_def` maybe.

nandojve replied:
I did all changes except name signature to **dt_str_from_compat**. I preferred keep the actual method name convention on the file since this was recently introduced. However, I don't have any problem to change it.

ulfalizer reviewed (commented):
Some nits and random thoughts.

Having some generic property lookup functions like these seems nice.

nandojve commented:
I assumed that **instance_no** will be enough to check but testing in deep I saw inconsistencies.
```
&foo0 {
	status = "disabled";
	bar@0 {
		compatible = "man,bar";
		reg = <0x0>;
		label = "WPAN_0";
		status = "disabled";
	};
	bar@1 {
		compatible = "man,bar";
		reg = <0x1>;
		label = "WPAN_1";
		status = "okay";
	};
};

&foo1 {
	status = "okay";
	bar@0 {
		compatible = "man,bar";
		reg = <0x0>;
		label = "RF2XX_2";
		status = "okay";
	};
}; 
```
On this example the first instance should be _/../foo1/bar@0_ and must be the only one. The current version returns _/../foo0/bar@0_. At **generated_dts_board.conf** file I saw two instances _/../foo0/bar@1_ and _/../foo1/bar@0_.

nandojve committed:
scripts: dts: edtlib.py: Fix DT enabled node property

The current version evaluate the node status and not the parent status.
That define unnecessary symbols at 'devicetree.conf' which can
increases total code size. This fix adding a new initialization method
that evaluate all parents from a node and disable when appropriated.

This fixes 'instance_no' set adding 'compat' only if node is enabled.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -823,8 +824,7 @@ def depends_on(self):
         "See the class docstring"
-        return "status" not in self._node.props or \
-            self._node.props["status"].to_string() != "disabled"
+        return self._node_status
```
Could remove this function and rename `Node._node_status` to `Node.enabled`.

nandojve replied:
Drop property and add public instance variable Node.enabled

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -890,6 +890,23 @@ def __repr__(self):
+        # evaluated by the node status property and If his parent status is
+        # equal to 'disabled' the node became 'disabled' too. This ensures that
+        # the whole tree branch is disabled.
+        self._node_status = "status" not in self._node.props or \
```
Would get `self.enabled = ...` here instead (and below).

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,45 @@ def dt_compat_enabled(kconf, _, compat):
+
+    try:
+        inst_no = int(inst)
+    except ValueError:
```
Could add a

```python
_warn(kconf, "malformed instance number")
```

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,45 @@ def dt_compat_enabled(kconf, _, compat):
+
+        if prop_s not in node.props:
+            break
+
```
Could do this just to be safe:

```python
val = node.props[prop_s].val

if not isinstance(val, (str, int)):
    _warn(kconf, "substituting devicetree property value that's not a string or an int")

return str(val)
```

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -890,6 +890,23 @@ def __repr__(self):
+            if not parent.enabled:
+                self._node_status = False
+                return
+            parent = parent.parent
```
The `for dt_node in self._dt.node_iter():` loop visits parent nodes before child nodes, so `self.parent.enabled` is already initialized here. Might work to replace `_init_status()` with this:

```python
parent_enabled = not self.parent or parent.enabled
self_enabled = "status" not in self._node.props or \
    self._node.props["status"] != "disabled"
return parent_enabled and self_enabled
```

nandojve replied:
Yes, no problem!

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1297,6 +1314,9 @@ def _set_instance_no(self):
 
+        if not self.enabled:
+            return
+
```
Does this make a difference for the output? `gen_defines.py` only seems to look at `node.instance_no` for enabled nodes.

nandojve replied:
Short history...the check above is necessary to keep `dt_compat_get_str` given same result of  `gen_defines.py`.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -890,6 +890,23 @@ def __repr__(self):
+        # Initializes Node._node_status.
+        #
+        # Node._node_status holds node enabled condition. This is dynamically
+        # evaluated by the node status property and If his parent status is
```
s/his/its/

Pronouns for dead things are never he/she in English, though I know where it's coming from. :)

nandojve replied:
Ops...text changed!

ulfalizer reviewed (changes requested):
Having `status = "disabled"` disable all the child nodes as well makes sense to me. Thanks for that.

I wonder how common it is to have nested `status` properties, but if they show up, it probably makes sense for the parent `status` to take precedence.

Feels like the old `_set_instance_no()` code could be cleaned up a bit. It's pretty subtle how it works (depends on `edt.nodes` only including the "previous" nodes when the function is called).

I'm thinking it might be nice to have an `edt.compat2nodes` property, where e.g. `edt.compat2nodes["foo"]` would be a list with all enabled nodes that have `compatible = ..., "foo", ...`. That would simplify `dt_compat_get_str()` as well.

I'll make a separate PR for that.

ulfalizer commented:
PR that turns `Node.instance_no` into `EDT.compat2nodes`: https://github.com/zephyrproject-rtos/zephyr/pull/21592

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -890,6 +885,17 @@ def __repr__(self):
+        #
+        # Evaluate 'status' property for both Node and Node.parent to
+        # determine final Node.enabled condition. This ensures that
+        # the whole tree branch is disabled when appropriated.
```
Small typo: *appropriate

nandojve replied:
:+1:

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -684,7 +685,7 @@ class Node:
 
     enabled:
-      True unless the node has 'status = "disabled"'
+      True when the full path have 'status = "okay"', False otherwise.
```
Could say

    True unless the node or one of its parents has 'status = "disabled"'`

Just to match the code.

nandojve replied:
:+1:

ulfalizer commented on drivers/ieee802154/Kconfig.rf2xx:
```diff
@@ -5,17 +5,20 @@
+		 label, \
+		 $(DT_IEEE802154_COMPAT_RF2XX), \
+		 0, \
+		 RF2XX_0)"
```
Could maybe put on one line and add a comment. Gets a bit over 80 chars, but no catastrophe.

Thinking something like this:

```
config NET_CONFIG_IEEE802154_DEV_NAME
	# Fetch 'label' property from instance 0, defaulting to "RF2XX_0"
	default "$(dt_compat_get_str,label,$(DT_IEEE802154_COMPAT_RF2XX),0,RF2XX_0)"
```

Maybe those comments would get annoying if it gets used in lots of places, but might be nice as long as it's just used in a few places.

nandojve replied:
yes, sure!

ulfalizer reviewed (approved):
Seems alright to me. Nice to have a function like that. Could be simplified a bit once https://github.com/zephyrproject-rtos/zephyr/pull/21592 is in too. Suspect most people are vacation at the moment.

Wonder if the function could just be called `dt_compat_get` too, since it also works for integers. Maybe there could be a `dt_compat_get_array` later if needed, that also takes an array index.

nandojve commented:
> Seems alright to me. Nice to have a function like that. Could be simplified a bit once #21592 is in too. Suspect most people are vacation at the moment.
>
> Wonder if the function could just be called `dt_compat_get` too, since it also works for integers. Maybe there could be a `dt_compat_get_array` later if needed, that also takes an array index.

I agreed with you, make sense.

nandojve commented:
Rebase on top of `dts: edtlib: Turn Node.instance_no into more flexible EDT.compat2nodes` #21592

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,49 @@ def dt_compat_enabled(kconf, _, compat):
+        return default_s
+
+    node_list = edt.compat2nodes[compat_s]
+    node = node_list[inst_no] if inst_no < len(node_list) else None
```
Could add a check for `compat_s` being in `edt.comp2nodes` as well:

```python
if compat_s not in edt.compat2nodes:
    return default_s

node_list = edt.compat2nodes[compat_s]
if inst_no >= node_list:
    return default_s

node = node_list[inst_no]
...
```

erwango commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -335,11 +335,52 @@ def dt_compat_enabled(kconf, _, compat):
     return "n"
 
 
+def dt_compat_get(kconf, _, prop, compat, inst, default=""):
```
I'd rename function to less generic `dt_compat_get_instance`.
We could think `dt_compat_get` could be useful and later used to retrieve all instances.

nandojve replied:
@ulfalizer gave me this explanation:
> Wonder if the function could just be called dt_compat_get too, since it also works for integers. Maybe there could be a dt_compat_get_array later if needed, that also takes an array index.

@erwango, I'm wonder if a function like **dt_compat_get_array** can address your concern?

erwango replied:
@nandojve, my point was on the nature of the result rather than its type.
Function seeks for one particular instance, seems logical to mention it, `dt_compat_get` by itself is misleading. 
Besides, since this function is intended to be used in Kconfig files, it is important that function name speaks by itself so reader does not need to reach function documentation.
Thinking about it, maybe `dt_compat_get_inst_prop` would be the best choice.

nandojve replied:
I did rebase and change from 'dt_compat_get' to 'dt_compat_get_inst_prop'.

tbursztyka reviewed (changes requested):
Mixing up DTS into Kconfig is, imo, a mistake.
After almost 3 years we have been using DTS, we are still not yet where we want to be (too many instance-centric options are still in Kconfig, etc..)
And that PR makes it worse by mixing both. Definitely not a good idea.

nandojve commented:
> Mixing up DTS into Kconfig is, imo, a mistake.
> After almost 3 years we have been using DTS, we are still not yet where we want to be (too many instance-centric options are still in Kconfig, etc..)
> And that PR makes it worse by mixing both. Definitely not a good idea.

The initial motivation of this PR was an open discussion at #21402. I agreed with you after weeks looking around DT, fixups, Kconfig etc. I believe any driver definition/configuration must be on that place and not around.

I would like mention that there are DT improvements on #21592 and bug fix at 'scripts: dts: edtlib.py: Fix DT enabled node property' that probably need move forward.

nandojve commented:
Created "scripts: dts: edtlib.py: Fix DT enabled node property" #22215 dependency to separate DT fix from the current PR.

ulfalizer commented:
> Mixing up DTS into Kconfig is, imo, a mistake.
> After almost 3 years we have been using DTS, we are still not yet where we want to be (too many instance-centric options are still in Kconfig, etc..)
> And that PR makes it worse by mixing both. Definitely not a good idea.

I think the problem is mostly Kconfig options that allow you to change stuff that's already known from selecting the board. 

Think of Kconfig like an options menu in a game. You wouldn't show an option for selecting the operating system there, since it's already known.

Having devicetree information available to Kconfig is something else though. It allows you to tweak the options based on based on what the devicetree says (e.g. to only show Windows-specific options if the game runs on Windows, just to extend that example).

This PR might help with getting rid of `devicetree.conf` too, which would be nice.

nandojve commented:
Since there isn't movement here I'm closing.
